text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+and there is currently no design of how blocks are to be retrieved in the Byzantine setting or how signatures on blocks should be verified.
+
+
+
+Block retrieval from Byzantine orderers needs to tackle denial of service from malicious orderers.
+
+The challenge with block signature validation in the Byzantine setting is avoiding the linear space overhead in the number of certificates that need to be stored as part of the signature information.
+
+While forks of Fabric with a Byzantine Fault Tolerant ordering service [exist](https://github.com/SmartBFT-Go/fabric/), they have not been incorporated into the official Hyperledger Fabric project.
+
+This RFC aims to:
+
+- Give an overview of how denial of service resistant block retrieval and block signature validation is done in SmartBFT.
+- Define how denial of service resistant block retrieval should be conducted in Hyperledger Fabric.
+- Define how Byzantine block signature validation should be constructed and encoded in Hyperledger Fabric.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Hyperledger Fabric only supports ordering service types that are Crash Fault Tolerant (CFT), not Byzantine Fault Tolerant (BFT).
+
+There are two main attacks that are considered outside of the crash fault tolerant thread model, but must be protected against in the Byzantine fault tolerant setting:
+
+**Block withholding**: A Byzantine orderer might be reluctant to send newly formed blocks to peers that connect to it, either because of malicious intent or to save egress network bandwidth charges from its hosting provider.
+ Fabric architecture relies on peers receiving newly formed blocks as fast as possible, because otherwise client transaction submission times out and smart contract simulation results in stale reads.
+
+**Block forgery**: A Byzantine orderer might try to forge a block and send it to peers or other ordering nodes that replicate from it, in order to alter the smart contract execution on peers by subterfuge.
+ One way of doing this, is to omit a transaction from a block, or re-order transactions within a block.
+ 
+To circumvent block forgery attacks, the following invariant should be preserved: *A set of signatures on a block should only be considered valid if it is impossible for the parties running the protocol to construct a set of signatures on a different block that is also considered valid*.
+
+The aforementioned invariant can be implemented in different ways, such as requiring a **threshold of signatures** from different orderers on a block, or requiring a single **threshold signature** on a block.
+The threshold itself can vary depending on the consensus algorithm and on the stage in it that the signatures is produced: Signing a block post consensus requires signatures from only `f+1` orderers (at the expense of latency), while signing during consensus requires signatures from `2f+1` orderers (at the expense of size and computation).
```

Comment:
For my own information and learning - could you point me to a reference that explains the difference here please? Why `f+1` and `2f+1`?

Reply:
It follows from the BFT protocol: If the BFT protocol has three stages, like PBFT, then the last stage involves collecting `2f+1` messages, each containing a hash on the block, and all `2f+1` messages need to be with the same hash. 

Since you want to transfer this block to someone else, that someone else needs to be sure that you really collected `2f+1` messages, it cannot take your word for it. So you make each message signed. Now when you send this block to someone else, it can verify the signatures and see that you really received the messages in the third phase from `2f+1` different nodes, so that block was produced in a sound execution of BFT.

But there is another way of collecting signatures that requires only `f+1` signatures. Consider a BFT protocol that doesn't use signatures at all, and after each block is committed, the nodes simply send each other signatures over the blocks they have agreed upon during the BFT protocol. If you have `f+1` signatures over such blocks, then it means that you have at least a single honest node (since up to `f` nodes are malicious) that participated in the BFT protocol and committed that block. From the correctness of the BFT protocol, it follows that it is the only possible block that can be agreed upon, and since you now know that an honest node committed that block, then you can also trust that block yourself.